bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -132,47 +133,51 @@ void ClientsManager::RequestsInfo::logAllPendingRequestsExceedingThreshold(const
 
 /*************************** Class ClientsManager::RepliesInfo ***************************/
 
-void ClientsManager::RepliesInfo::deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient) {
-  Time earliestTime = MaxTime;
-  ReqId earliestReplyId = 0;
-  if (repliesMap_.size() < maxNumOfReqsPerClient) return;
-  if (repliesMap_.size() > maxNumOfReqsPerClient)
+void ClientsManager::RepliesInfo::deleteReplyIfNeededSafe(NodeIdType clientId,
+                                                          ReqId reqSeqNum,
+                                                          uint16_t maxNumOfReqsPerClient,
+                                                          uint16_t reqIndex) {
+  ReqId savedReplyId = 0;
+  if (repliesBiMap_.size() > maxNumOfReqsPerClient) {
     LOG_FATAL(CL_MNGR,
               "More than maxNumOfReqsPerClient_ items in repliesInfo"
-                  << KVLOG(repliesMap_.size(), maxNumOfReqsPerClient, clientId));
-  for (auto& reply : repliesMap_) {
-    if (earliestTime > reply.second) {
-      earliestReplyId = reply.first;
-      earliestTime = reply.second;
+                  << KVLOG(repliesBiMap_.size(), maxNumOfReqsPerClient, clientId));
+  }
+
+  if (repliesBiMap_.right.find(reqIndex) != repliesBiMap_.right.end()) {
+    savedReplyId = repliesBiMap_.right.at(reqIndex);
+    if (savedReplyId > reqSeqNum) {
+      LOG_WARN(CL_MNGR, "A newer reply was already saved" << KVLOG(clientId, reqSeqNum, savedReplyId, reqIndex));
     }
   }
-  const lock_guard<mutex> lock(repliesMapMutex_);
-  if (earliestReplyId) {
-    repliesMap_.erase(earliestReplyId);
-  } else if (!repliesMap_.empty()) {
-    // Delete reply arrived through ST
-    auto const& reply = repliesMap_.cbegin();
-    earliestReplyId = reply->first;
-    earliestTime = reply->second;
-    repliesMap_.erase(reply);
+
+  if (savedReplyId) {
+    // The thread safety here is between the PreProcessor thread, which is read-only, and the ReplicaImp main thread,
+    // which performs either read or write at a time. So, all write operations should be guarded as well as PrePropessor
+    // calls hasReply and isClientRequestInProcess. All other operations are safe.
```

Comment:
I would move this comment to the hpp file and put it before 'safe' functions of the RepliesInfo class.